[Threading][TSan] Fix TSan errors from lazy init on Linux.

al45tair opened the pull request:
Move the TSan functionality from Concurrency into Threading.  Use it in the Linux `ulock` implementation so that TSan knows about `ulock` and will tolerate the newer `swift_once` implementation that uses it.

This should also slightly improve the performance of Concurrency, since the tsan tests will be inlined rather than it always doing a subroutine call.

rdar://110665213

al45tair committed:
[Threading][TSan] Fix TSan errors from lazy init on Linux.

Move the TSan functionality from Concurrency into Threading.  Use it
in the Linux `ulock` implementation so that TSan knows about `ulock`
and will tolerate the newer `swift_once` implementation that uses it.

rdar://110665213

al45tair commented:
@swift-ci Please test

yln commented on include/swift/Threading/TSan.h:
```diff
@@ -0,0 +1,95 @@
+  return threading_impl::tsan_enabled;
+}
+
+/// Indicate to TSan that an acquiring load has occurred on the current
```
Same as my other comment.  Not sure about the wording to explain the APIs here.

Example: T1 & T2 contending on the same lock `x` (or using some other API to "synchronize" / establish order between them):
```
T1                | T2

write to y
tsan_release(x)
lock given away
      --> sync point -->
                     lock taken
                     tsan_aquire(x)
                     read from y
```

TSan now knows: T1 release(x) synchronized/happens before T2 acquire(x), access to y is well-ordered, don't diagnose race for accesses to y

al45tair replied:
Having read this again this morning, I think we're describing the exact same thing; the point is that all writes that happened-before the `tsan_release(x)` are visible to the thread that did `tsan_acquire(x)`, thus the read is safe. I think the only difference is that the comments I added are describing things from the perspective of the C memory model rather than from the perspective of the internal implementation of TSan, but the fundamental point is that you need a `tsan_acquire()` to correspond to an acquiring load that TSan didn't see, and a `tsan_release()` to correspond to a releasing store that TSan didn't see.

The difference between _acquire_ and _consume_ has to do with carried dependencies; OTOH consume is weaker than acquire, the point being that you could use it to protect a structure to which you had a pointer, without forcing memory ordering constraints on anything that didn't involve an access through that pointer. Modelling it as a `tsan_acquire()` is kind of wrong, since on a CPU that does support consume, non-dependent reads of data written before the corresponding `tsan_release()` would technically be a bug.

yln commented on include/swift/Threading/TSan.h:
```diff
@@ -0,0 +1,95 @@
+/// acquire instead.
+template <typename T>
+T *consume(T *ptr) {
+  return acquire(ptr);
```
Please skip adding `tsan::consume()` for now.

al45tair replied:
OK, though I only added that because the code in Concurrency had it and really I'm just pushing that down into the Threading library. I can make Concurrency's version call `tsan::acquire()` instead, but I don't want to remove it there because (presumably) Concurrency is using it somehow.

al45tair replied:
Returning to this this morning, I think it's probably best to not add `tsan::consume` now as well. Calling `_tsan_acquire` for _consume_ is potentially problematic, since TSan won't detect non-dependent accesses and that could lead to a false sense of security.

yln commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -99,15 +100,18 @@ void _swift_taskGroup_cancelAllChildren(TaskGroup *group);
+}
+/// Tell TSAN about a consuming load
+inline void _swift_tsan_consume(void *addr) {
+  swift::tsan::consume(addr);
```
Can be implemented by delegating to `_swift_tsan_acquire` or stay as a macro to avoid the need for consume until I caught up with why this is needed.

yln commented on test/Sanitizers/tsan/once.swift:
```diff
@@ -0,0 +1,50 @@
+// Test that we do not report a race on initialization; Swift doesn't initialize
+// globals at start-up, but rather uses swift_once().  This is thread safe, but
+// on some platforms TSan wasn't seeing the synchronization, so would report
+// a false positive.
```
Please extend this comment to explain how this is testing `swift_once()` (Swift internal!) to avoid confusing, since the test code itself doesn't mention `once()`.

al45tair replied:
Good point. The actual original symptoms of the problem are that code like the code in the test was triggering TSan messages, so from that perspective `swift_once()` is also an implementation detail, but expanding the comment seems good.

al45tair commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -99,15 +100,18 @@ void _swift_taskGroup_cancelAllChildren(TaskGroup *group);
-/// reason about this, we tell TSAN it's an acquire instead. See also
-/// SWIFT_MEMORY_ORDER_CONSUME definition.
-#define _swift_tsan_consume _swift_tsan_acquire
+/// Tell TSAN about an acquiring load
```
:-) I was just trying to expand on the original comments from the functions in Concurrency, which seemed (to me) to be referring to the meanings from the C/C++ memory model. If acquire/release/consume doesn't mean the same as it does in C/C++, then perhaps the terminology was ill-advised (though we're stuck with it now); I'm quite happy to update the comments to say whatever you think they should.

al45tair replied:
(See below for more).

al45tair commented on include/swift/Threading/Impl/Linux/ulock.h:
```diff
@@ -59,32 +61,39 @@ inline void ulock_lock(ulock_t *lock) {
+      break;
   } while (ulock_futex(lock, FUTEX_UNLOCK_PI) != 0);
+
+  tsan::release(lock);
```
:-) I did that initially, but then changed it in favour of uniformity in the implementation — I didn't think it'd make any difference in this case because TSan can't see what's happening inside the runtime (so whether from its perspective the locking takes place before or after isn't that relevant). But maybe you're right — I'll move it back again.

finagolfin commented:
Note that your new tsan test is not run on the linux CI, most likely because of #53973:
```
UNSUPPORTED: Swift(linux-x86_64) :: Sanitizers/tsan/once.swift (6673 of 16678)
```

al45tair committed:
[Threading][TSan] Update after review comments.

* Use the longer name ThreadSanitizer rather than TSan for the new files.
* Don't implement `tsan::consume` at all for now.
* Do the `tsan::release` for `ulock_unlock()` at the head of the function,
  not at the tail.
* Add a comment to test/Sanitizers/tsan/once.swift to explain the test a
  little more clearly.

rdar://110665213

al45tair commented:
@swift-ci Please smoke test

weissi commented on include/swift/Threading/Impl/Linux/ulock.h:
```diff
@@ -59,31 +61,38 @@ inline void ulock_lock(ulock_t *lock) {
+    if (ulock_fastpath(__atomic_compare_exchange_n(lock, &zero, tid,
+                                                   true, __ATOMIC_ACQUIRE,
+                                                   __ATOMIC_RELAXED)))
+      break;
```
:D

weissi reviewed (approved):
Lovely, thanks so much!

al45tair commented:
OK, so this is broken on macOS because of the behaviour of `RTLD_NEXT` there. Might need to rearrange things a little.

al45tair committed:
[Threading][TSan] Move ThreadSanitizer.cpp into the main runtime.

On Darwin, `RTLD_NEXT` doesn't do what we need it to here, with the
result that if `libswiftCore`'s TSan initializer gets found first,
then `libswift_Concurrency` won't have its initializer called at all,
in spite of us using `RTLD_NEXT` to find the next definition.

Fix this by centralising the initializer in `libswiftCore` instead.

rdar://110665213

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
(Rebased to fix conflict)

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
I think this version will likely fail the unit tests on Linux, because various things besides the runtime libraries use the Threading library, and not all of them link against the runtime. It looks like I'm going to need to rearrange things a little more.

al45tair committed:
[Threading][TSan] Rearrange things again.

We need ThreadSanitizer.cpp in libswiftCore for the runtime case, but
we also need it in libswiftThreading for non-runtime cases.

rdar://1106655213

al45tair committed:
[Threading][TSan] Fix linkage issue.

We need to pick up the `_swift_tsan_xxx` symbols from libswiftCore in
most cases, but sometimes we're statically linked and in that case we
want to use a local copy.

rdar://1106655213

al45tair committed:
[Threading][TSan] Fix builds where TSan isn't supported.

On builds where TSan isn't supported, don't include any code in
ThreadSanitizer.cpp.

rdar://110665213

al45tair commented:
@swift-ci Please smoke test

yln reviewed (approved):
Thanks so much Alastair!

LGTM, with one nit.

Please document the annotation APIs in terms of TSan "model synchronization operation" instead of C memory model.

Something like we had before:
```
/// release() establishes a happens-before relation with a preceding acquire()
/// on the same address.
```

Something that describes that we are modeling / "teaching" TSan about asynchronization operation (not necessarily lock-based):
```
T1                | T2

access from y
tsan_release(x)
lock given away
      --> sync point -->     <<< this is what we try to model
                     lock taken
                     tsan_aquire(x)
                     access to y
```

al45tair committed:
[Threading][TSan] Update comments.

Updated the comments for tsan::acquire and tsan::release to better
reflect what TSan is actually doing.

rdar://110665213

al45tair commented:
@swift-ci Please smoke test
